pkg/scheduler/metrics/resources/resources.go

```diff
@@ -0,0 +1,296 @@
+// reports the current resources requested by all pods on the cluster within
+// the Kubernetes resource model. Metrics are broken down by pod, node, resource,
+// and phase of lifecycle. Each pod returns two series per resource - one for
+// their aggregate usage (required to schedule) and one for their phase specific
+// usage. This allows admins to assess the cost per resource at different phases
+// of startup and compare to actual resource usage.
+func NewPodResourcesMetricsCollector(podLister corelisters.PodLister) metrics.StableCollector {
+	return &podResourceCollector{
+		lister: podLister,
+	}
+}
+
+type podResourceCollector struct {
+	metrics.BaseStableCollector
+	lister corelisters.PodLister
+}
+
+func (c *podResourceCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	podResourceDesc.Describe(ch)
+}
+
+func (c *podResourceCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	pods, err := c.lister.List(labels.Everything())
+	if err != nil {
+		return
+	}
+	reuseReqs, reuseLimits := make(v1.ResourceList, 4), make(v1.ResourceList, 4)
+	for _, p := range pods {
+		_, reqs, _, limits, _, terminal := podRequestsAndLimitsByLifecycle(p, reuseReqs, reuseLimits)
+		if terminal {
+			// terminal pods are excluded from resource usage calculations
+			continue
+		}
+		for _, t := range []struct {
+			desc    resourceLifecycleDescriptors
+			total   v1.ResourceList
+			running v1.ResourceList
+		}{
+			{
+				desc:  podResourceDesc.requests,
+				total: reqs,
+			},
+			{
+				desc:  podResourceDesc.limits,
+				total: limits,
+			},
+		} {
+			for resourceName, val := range t.total {
+				var unitName string
+				switch resourceName {
+				case v1.ResourceCPU:
+					unitName = "cores"
+				case v1.ResourceMemory:
+					unitName = "bytes"
+				case v1.ResourceStorage:
+					unitName = "bytes"
+				case v1.ResourceEphemeralStorage:
+					unitName = "bytes"
+				default:
+					switch {
+					case isHugePageResourceName(resourceName):
+						unitName = "bytes"
+					case isAttachableVolumeResourceName(resourceName):
+						unitName = "integer"
+					}
+				}
+				var priority string
+				if p.Spec.Priority != nil {
+					priority = strconv.FormatInt(int64(*p.Spec.Priority), 10)
+				}
+				recordMetricWithUnit(ch, t.desc.total, resourceName, unitName, val, p.Namespace, p.Name, p.Spec.NodeName, p.Spec.SchedulerName, priority)
+			}
+		}
+	}
+}
+
+func recordMetricWithUnit(
+	ch chan<- metrics.Metric,
+	desc *metrics.Desc,
+	resourceName v1.ResourceName,
+	unit string,
+	val resource.Quantity,
+	namespace, name, nodeName, schedulerName, priority string,
+) {
+	if val.IsZero() {
+		return
+	}
+	ch <- metrics.NewLazyConstMetric(desc, metrics.GaugeValue,
+		val.AsApproximateFloat64(),
+		namespace, name, nodeName, schedulerName, priority, string(resourceName), unit,
+	)
+}
+
+// addResourceList adds the resources in newList to list
+func addResourceList(list, newList v1.ResourceList) {
```

Comment:
So this is reimplementing the logic again, which is against the purpose of putting this in the scheduler code.

Here is one implementation: https://github.com/kubernetes/kubernetes/blob/d9d626f1e67f2a6eb1d68ed06b265fa877ec5764/pkg/scheduler/framework/plugins/noderesources/fit.go#L148:6

And here is another:
https://github.com/kubernetes/kubernetes/blob/d9d626f1e67f2a6eb1d68ed06b265fa877ec5764/pkg/scheduler/framework/types.go#L601:6

I'm not sure if you will be able to use either, but at least at a TODO for the unification of the code.
Here is the cleanup bug that was not worked on 😞  #90402

Reply:
Trivial helper reuse is not something I consider reimplementation.  Agree on some of the other types below.

Reply:
I'm not opposed to moving these to helpers though, although if we do helpers i would like them in staging rather than in k/k

Reply:
It might be trivial, but the duplication has already lead to bugs. Please add the TODO and link to cleanup issue.

Reply:
Within the context of the scheduler I agree on better reuse (your examples above).  I'm less supportive of reuse across broad domains (scheduler / kubelet) unless it's a standard API, and even then there has to be some caution because I've seen this mask other problems and changing that behavior then has client impacts.